src/core/Do.Architecture/Architecture/Application.cs

```diff
@@ -12,14 +12,27 @@ public class Application
 
     internal Application With(ApplicationDescriptor descriptor)
     {
+        CheckDuplicates(descriptor.Layers.Select(layer => layer.Id).ToList());
         _layers.AddRange(descriptor.Layers);
+
+        CheckDuplicates(descriptor.Features.Select(feature => feature.Id).ToList());
         _features.AddRange(descriptor.Features);
 
         FillPhases();
 
         return this;
     }
 
+    void CheckDuplicates(List<string> ids)
+    {
+        var duplicates = ids.GroupBy(x => x).Where(g => g.Count() > 1).Select(l => l.Key).ToList();
+
+        if (duplicates.Any())
+        {
+            throw new InvalidOperationException($"Cannot add `{string.Join(", ", duplicates)}`, it was already added.");
```

Comment:
you include all duplicates, but you don't test if you include them or not. you are adding code you don't test. your tests and code should be consistent.

just include the first duplicate in exception message.